Cython/Utility/TypeConversion.c

```diff
@@ -241,6 +241,31 @@ static CYTHON_INLINE const char* __Pyx_PyObject_AsString(PyObject* o) {
 #if __PYX_DEFAULT_STRING_ENCODING_IS_ASCII || __PYX_DEFAULT_STRING_ENCODING_IS_DEFAULT
 static CYTHON_INLINE const char* __Pyx_PyUnicode_AsStringAndSize(PyObject* o, Py_ssize_t *length) {
     if (unlikely(__Pyx_PyUnicode_READY(o) == -1)) return NULL;
+#if CYTHON_COMPILING_IN_LIMITED_API && __PYX_LIMITED_VERSION_HEX < 0x030A0000
+    PyErr_SetString(
```

Comment:
As far as I can tell, `__Pyx_PyUnicode_AsStringAndSize` is only ever used in `__Pyx_PyObject_AsStringAndSize` so it's always used on something that might be a unicode object but isn't known to be. Therefore a runtime error seems appropriate. If it was definitely used on a unicode object I'd probably prefer a compile-time error.